### ISSUE TITLE
"implementation-defined"

### ISSUE THREAD

**Issue by annevk** [whatwg/infra#210]
Maybe we should define "implementation-defined" as a term to use when it's really up to the implementation. At the same time we can stipulate that unless that phrase is used, implementations have basically no freedom. (Except UI and such I suppose, I guess I need to double check how that's normally handled.)

**Comment by jugglinmike** [comment]
Recently, I came across [a case where Firefox dispatches a DOM event in the absence of any normative text mandating it](https://github.com/whatwg/html/issues/1518#issuecomment-410118468). @annevk mentioned me above because I expressed uncertainty about whether it was appropriate to write a test for web-platform-tests to verify that no such event is dispatched. As Anne's summary makes clear, such a test *is* valid.

A formal definition for "implementation-defined" would help test authors like me discover the notion and learn the general policy about arbitrary extensions.  That may seem obvious, but folks coming from ECMA262 (where [implementations are free to add new APIs and even new syntax](https://tc39.github.io/ecma262/#sec-conformance)) might be surprised by WHATWG's strictness.
